test/Examples.hs

```diff
@@ -105,8 +68,57 @@ languages =
   -- , ("php", ".php") -- TODO: No parse-examples in tree-sitter yet
   ] where examples = Path.relDir "examples"
 
+buildExamples :: TaskSession -> LanguageExample -> Path.RelDir -> IO Tasty.TestTree
+buildExamples session lang tsDir = do
+  knownFailures <- knownFailuresForPath tsDir (languageKnownFailuresTxt lang)
+  files <- globDir1 (compile ("**/*" <> languageExtension lang)) (Path.toString (tsDir </> languageExampleDir lang))
+  let paths = Path.relFile <$> files
+  trees <- for paths $ \file -> pure $ HUnit.testCase (Path.toString file) $ do
+    res <- runTask session (parseFilePath file)
+    case res of
+      Left (SomeException e) -> case cast e of
+        -- We have a number of known assignment timeouts, consider these pending specs instead of failing the build.
+        Just AssignmentTimedOut -> pure ()
+        Just ParserTimedOut     ->     pure ()
+        -- Other exceptions are true failures
+        _                       -> HUnit.assertFailure (show (displayException e))
+      _ -> if file `elem` knownFailures
+              then pure ()
+              else (isRight res) HUnit.@? ("Error: " <> either show show res)
+  pure (Tasty.testGroup (languageName lang) trees)
+
+testOptions :: Config.Options
+testOptions = defaultOptions
+  { optionsFailOnWarning = flag FailOnWarning True
+  , optionsLogLevel = Nothing
+  }
+
+main :: IO ()
+main = withOptions testOptions $ \ config logger statter -> do
+  void $ Process.system "script/clone-example-repos"
+
+  let session = TaskSession config "-" False logger statter
+
+  allTests <- forConcurrently examples $ \lang@LanguageExample{..} -> do
+    let tsLang = Path.relDir ("tree-sitter-" <> languageName)
+    let tsDir = Path.relDir "tmp/haskell-tree-sitter" </> tsLang </> Path.relDir "vendor" </> tsLang
+    buildExamples session lang tsDir
+
+  Tasty.defaultMain $ Tasty.testGroup "parse-examples" allTests
+
+knownFailuresForPath :: Path.RelDir -> Maybe Path.RelFile -> IO (Set Path.RelFile)
+knownFailuresForPath _ Nothing = pure mempty
+knownFailuresForPath tsDir (Just path)
+  = runResourceT
+  ( ByteStream.readFile @ResIO (Path.toString (tsDir </> path))
+  & ByteStream.lines
+  & ByteStream.denull
+  & Stream.mapped ByteStream.toLazy
+  & Stream.filter ((/= '#') . BLC.head)
+  & Stream.map (Path.relFile . BLC.unpack)
+  & Foldl.purely Stream.fold_ Foldl.set
```

Comment:
How’s the time/space perf relative to `master`?

Reply:
It should be improved, but this isn’t the bottleneck. I just like to turn uses of `lines <$> readFile` into streaming pipelines, since that’s usually the right thing to do and it’s generally just as concise.